stored_requests/config/config.go

```diff
@@ -22,48 +22,152 @@ import (
 // In the future we should look for ways to simplify this so that it's not doing two things.
-func NewStoredRequests(cfg *config.Configuration, client *http.Client, router *httprouter.Router) (fetcher stored_requests.Fetcher, ampFetcher stored_requests.Fetcher, db *sql.DB, shutdown func(), categoriesFetcher stored_requests.CategoryFetcher) {
-	if cfg.StoredRequests.Postgres.ConnectionInfo.Database != "" {
-		glog.Infof("Connecting to Postgres for Stored Requests. DB=%s, host=%s, port=%d, user=%s",
-			cfg.StoredRequests.Postgres.ConnectionInfo.Database,
-			cfg.StoredRequests.Postgres.ConnectionInfo.Host,
-			cfg.StoredRequests.Postgres.ConnectionInfo.Port,
-			cfg.StoredRequests.Postgres.ConnectionInfo.Username)
-		db = newPostgresDB(cfg.StoredRequests.Postgres.ConnectionInfo)
-	}
-	eventProducers, ampEventProducers := newEventProducers(&cfg.StoredRequests, client, db, router)
-	cache := newCache(&cfg.StoredRequests)
-	ampCache := newCache(&cfg.StoredRequests)
-	fetcher, ampFetcher = newFetchers(&cfg.StoredRequests, client, db)
-	categoriesFetcher, _ = newFetchers(&cfg.CategoryMapping, client, db)
-
-	fetcher = stored_requests.WithCache(fetcher, cache)
-	ampFetcher = stored_requests.WithCache(ampFetcher, ampCache)
-
-	shutdown1 := addListeners(cache, eventProducers)
-	shutdown2 := addListeners(ampCache, ampEventProducers)
+func CreateStoredRequests(cfg *config.StoredRequestsSlim, client *http.Client, router *httprouter.Router) (fetcher stored_requests.AllFetcher, db *sql.DB, shutdown func()) {
+	// Create database connection if given options for one
+	if cfg.Postgres.ConnectionInfo.Database != "" {
+		conn := cfg.Postgres.ConnectionInfo.ConnString()
+
+		if dbConnection.conn == "" {
+			glog.Infof("Connecting to Postgres for Stored Requests. DB=%s, host=%s, port=%d, user=%s",
+				cfg.Postgres.ConnectionInfo.Database,
+				cfg.Postgres.ConnectionInfo.Host,
+				cfg.Postgres.ConnectionInfo.Port,
+				cfg.Postgres.ConnectionInfo.Username)
+			db = newPostgresDB(cfg.Postgres.ConnectionInfo)
+			dbConnection.conn = conn
+			dbConnection.db = db
+		}
+
+		// Error out if config is trying to use multiple database connections for different stored requests (not supported yet)
+		if conn != dbConnection.conn {
+			glog.Fatal("Multiple database connection settings found in Stored Requests config, only a single database connection is currently supported.")
+		}
+	}
+
+	eventProducers := newEventProducers(cfg, client, db, router)
+	fetcher = newFetcher(cfg, client, db)
+
+	var shutdown1 func()
+
+	if cfg.InMemoryCache.Type != "" {
+		cache := newCache(cfg)
+		fetcher = stored_requests.WithCache(fetcher, cache)
+		shutdown1 = addListeners(cache, eventProducers)
+	}
+
 	shutdown = func() {
-		shutdown1()
-		shutdown2()
-		if db != nil {
+		if shutdown1 != nil {
+			shutdown1()
+		}
+		if dbConnection.db != nil {
+			db := dbConnection.db
+			dbConnection.db = nil
+			dbConnection.conn = ""
```

Comment:
Maybe the above two lines should execute only if/after the db.Close() below is confirmed to be successful!?